lib/pure/complex.nim

```diff
@@ -9,78 +9,150 @@
 
 
 
+
+
 ## This module implements complex numbers.
-{.push checks:off, line_dir:off, stack_trace:off, debugger:off.}
-# the user does not want to trace a part
-# of the standard library!
+## Complex numbers are currently implemented as generic on a 64-bit or 32-bit float.
 
+{.push checks: off, line_dir: off, stack_trace: off, debugger: off.}
+# the user does not want to trace a part of the standard library!
 
-import
-  math
 
-const
-  EPS = 1.0e-7 ## Epsilon used for float comparisons.
+import math
 
 type
-  Complex* = tuple[re, im: float]
-    ## a complex number, consisting of a real and an imaginary part
+  Complex*[T: SomeFloat] = tuple[re, im: T]
+    ## A complex number, consisting of a real and an imaginary part.
+  Complex128* = Complex[float64]
+    ## Alias for a pair of 64-bit floats.
+  Complex64* = Complex[float32]
+    ## Alias for a pair of 32-bit floats.
+
 
 const
-  im*: Complex = (re: 0.0, im: 1.0)
+  im*: Complex[float] = (re: 0.0, im: 1.0)
     ## The imaginary unit. √-1.
 
-proc toComplex*(x: SomeInteger): Complex =
-  ## Convert some integer ``x`` to a complex number.
-  result.re = x
-  result.im = 0
 
-proc `==` *(x, y: Complex): bool =
-  ## Compare two complex numbers `x` and `y` for equality.
+proc toComplex*[T: SomeNumber](x: T): auto =
+  ## Convert ``x`` to Complex.
+  when T is SomeInteger:
+    result.re = float(x)
+    result.im = 0.0
+  elif T is float32:
+    var z: Complex[float32]
+    z.re = x
+    z.im = float32(0)
+    result = z
+  elif T is float64:
+    var z: Complex[float64]
+    z.re = x
+    z.im = float64(0)
+    result = z
+
+proc toComplex64*[T: SomeNumber](x: T): Complex64 =
+  ## Convert ``x`` to Complex64.
+  result.re = float32(x)
+  result.im = float32(0)
+
+proc toComplex128*[T: SomeNumber](x: T): Complex128 =
+  ## Convert ``x`` to ``Complex128``.
+  result.re = float64(x)
+  result.im = float64(0)
+
+
+converter toComplex128*[T: Complex](z: T): Complex128 =
+  ## Convert some complex number ``z`` to a 128-bit complex number.
+  let (re, im) = z
+  result.re = re.float64
+  result.im = im.float64
+
+converter toComplex64*[T: Complex](z: T): Complex64 =
```

Comment:
converter with potential precision loss are not desirable. You should constrain that to literals

`converter toComplex64*[T: Complex](z: T{lit}): Complex64 =`

Also there is no automatic float32->float64 in Nim and this helps spots bugs.